docs: migrate 'Build and staging dependencies' (#5452)
canonical/snapcraft: docs/how-to/crafting/manage-dependencies.rst

Manage dependencies
===================

When building a snap and constructing a :ref:`part <explanation-parts>`, you commonly
need to specify build and staging dependencies. Build dependencies are required for your
part to successfully build or compile on the development host, while staging
dependencies are required for your snap to run.

For further help on solving build dependencies, see
:ref:`iterate-on-the-build-lifecycle`.

> Define dependencies
> -------------------

Reviewer: ```suggestion
Define a dependency
-------------------
```

Build and staging dependencies are added to a snap's :ref:`project file
<reference-snapcraft-yaml>` within a part definition. They can be added as standard

> packages for the chosen build environment, such as deb packages for Ubuntu, or as a

Reviewer: @mr-cal Debian doesn't officially call them "deb packages", but instead simply "packages" or "Debian packages". 

Formatting it as a file extension (`.deb`) every time is cumbersome. What do you think about formalising "Debian packages" as our preferred term?

Reviewer: After discussion, we agreed to go with "Debian package".

Author: https://github.com/canonical/snapcraft/pull/5452/commits/7c28fa2e46d8b5c4f25fbf2fb440dbece41ac8c1

cross-platform snap using the following keys:

.. list-table::
    :header-rows: 1

    * - Key
      - Description
    * - ``build-packages``
      - packages required for the part to build
    * - ``stage-packages``
      - packages required to run the part
    * - ``build-snaps``
      - snaps required for the part to build
    * - ``stage-snaps``
      - snaps required to run the part

The following is a typical example of a part's ``build-`` and ``stage-`` sections for a

> command-line tool that interacts with git:

Reviewer: ```suggestion
snap of a command-line tool that interacts with Git:
```

.. code-block:: yaml
    :caption: snapcraft.yaml

    build-packages:
      - pkg-config
      - libreadline-dev
      - libncurses5-dev
    build-snaps:
      - go
    stage-snaps:
      - ffmpeg/latest/edge
    stage-packages:
      - git

> Snaps are downloaded from the `Snap Store <https://snapcraft.io/store>`_ and unpacked

Reviewer 2: ```suggestion
Staged snaps are downloaded from the `Snap Store <https://snapcraft.io/store>`_ and unpacked
```

Author: https://github.com/canonical/snapcraft/pull/5452/commits/a2666f15d6a1801dee19176216a31e81d1da61e0

unpacked into the snap being built.

For staged snaps, the ``meta`` and ``snap`` directories from the snap will be available
as ``meta.<snap-name>`` and ``snap.<snap-name>`` for cases where assets from those
locations are desired for reuse.

> Package and snap names

Reviewer: This feels like it belongs before _Define a dependency_.

> Package and snap names
> ----------------------

Reviewer: ```suggestion
Identify a dependency
---------------------
```

Package dependencies are listed as package names for the snap's build environment.

> For a default Snapcraft installation running `Multipass <https://multipass.run/>`_ or
> `LXD <https://linuxcontainers.org/lxd/introduction/>`_, the build environment is
> dependent on whatever :ref:`base snap <base-snap-reference>` is being used.

Reviewer 2: This feels awkward but I can't figure out how to reword it. I think it may be simpler to drop it and assume Snapcraft is being used "normally".

```suggestion
Snapcraft's build environment is dependent on the :ref:`base snap <base-snap-reference>` used in the project file.
```

Author: I thought this passage was a bit unnatural as well. I trimmed it down as you suggested. :+1: 

https://github.com/canonical/snapcraft/pull/5452/commits/a2666f15d6a1801dee19176216a31e81d1da61e0

> :ref:`reference-build-environment-options` for further details

Reviewer: ```suggestion
:ref:`reference-build-environment-options` for further details.
```

> Consequently, dependencies are listed using their respective apt package names. See
> :ref:`reference-build-environment-options` for further details

Reviewer 2: ```suggestion
Consequently, dependencies are listed using their respective apt package names from the `Ubuntu package archive <https://packages.ubuntu.com/>`_. See
:ref:`reference-build-environment-options` for further details
```

Author: https://github.com/canonical/snapcraft/pull/5452/commits/a2666f15d6a1801dee19176216a31e81d1da61e0

It's also feasible to have a build environment built on Fedora for example, using Fedora
packages, or those of your own host environment.

The required packages are likely to be identical to those needed to either build the
project (``build-packages``) or install your project (``stage-packages``) natively.
You'll often find them listed in a project's README, or alongside any build
instructions.

Snap names are identical to the names used to install the snap outside of the build
environment, and those listed by the Snap Store. If a specific channel is required, the
syntax is of the form ``<snap-name>/<channel>`` (see :ref:`reference-channels` for more

> details on the syntax).

Reviewer: @jahn-junior This is more elementary information than the first section. What do you think about making this first?

Author: https://github.com/canonical/snapcraft/pull/5452/commits/7c28fa2e46d8b5c4f25fbf2fb440dbece41ac8c1

> Identify missing packages
> -------------------------

Reviewer: ```suggestion
Resolve missing dependencies
----------------------------
```

Working out your project's dependencies can be an iterative process, much like compiling
a third-party package, with the process split into identifying the dependencies a snap

> needs for a successful build, and those required for running the resultant application.

Reviewer: ```suggestion
needs for a successful build, and those required for running the resulting application.
```

needs for a successful build, and those required for running the resultant application.

Missing build dependencies
~~~~~~~~~~~~~~~~~~~~~~~~~~

The following steps outline a basic process for identifying missing packages in your
snap:

#. Once your snap has a functional project file, run ``snapcraft --debug`` until you
   encounter an error.
#. If the error is a build dependency, use the interactive debug shell to work out
   which package is required.
#. Add the package to your project file and repack the snap to see if you encounter
   the same error.

A typical missing build dependency will generate an error similar to the following:

.. terminal::

    configure: error: can't find the Boehm GC library.  Please install it.
    Failed to run 'override-build': Exit code was 1.

In most cases, the error will provide some indication of what needs to be installed. To
resolve the error from the previous example output, the Boehm GC library (``libgc``)
will need to be installed, and because this is the building stage, so too will its
header package (``libgc-dev``).

> If the package is unknown, its correct name can normally be found with a search from
> within the build environment, or via a search engine.

Reviewer 2: ```suggestion
If the package is unknown, its correct name can normally be found with a search from
within the build environment, a search engine, or on the `Ubuntu package archive <https://packages.ubuntu.com/>`_.
```

Author: https://github.com/canonical/snapcraft/pull/5452/commits/a2666f15d6a1801dee19176216a31e81d1da61e0

> within the build environment, a search engine, or on the `Ubuntu package archive
> <https://packages.ubuntu.com/>`_.

Reviewer: ```suggestion
within the build environment, on the `Ubuntu package archive
<https://packages.ubuntu.com/>`_, or searching online.
```

Missing staging dependencies
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> After a snap has been successfully built, if Snapcraft detects that the resultant

Reviewer: ```suggestion
After a snap has been successfully built, if Snapcraft detects that the resulting
```

running its executable will result in an error. The reasons for these errors are varied,
but the most common is a missing library, as shown in the following example output:

.. terminal::

    /snap/mysnap/current/bin/mybin: error while loading shared libraries: libpaho-mqtt3a.so.1: cannot open shared object file: No such file or directory

The following are the most common solutions for these kinds of errors:

* If a required library might not have been installed by the snap, add the missing
  package to the appropriate part's ``stage-packages`` key.
* If the snap's ``LD_LIBRARY_PATH`` environment variable doesn't include the path to the
  missing library, update ``LD_LIBRARY_PATH`` in the snap's project file. The following
  example adds ``$SNAP/usr/lib``:

.. code-block:: yaml
    :caption: snapcraft.yaml

    apps:
      example-app:
        [...]
        environment:
          LD_LIBRARY_PATH: $LD_LIBRARY_PATH:$SNAP/usr/lib

* If the missing library is shared to the snap by a content interface that isn't
  properly connected, connect the snapd interface. This should be done manually for
  testing and automatically for production.

> Remove stage package duplication
> --------------------------------

Reviewer: @mr-cal I'd like your opinion on this:

```suggestion
Override a core dependency
--------------------------
```

Reviewer 2: Oh, good catch.  I like your suggestion.

Reviewer: [36d7fc3](https://github.com/canonical/snapcraft/pull/5452/commits/36d7fc39053fd35714292057809def4f57ce376b)
